### PR TITLE
[Inductor][CPU] Fix int8/uint8 min/max vectorized reductions polluted by inactive lanes

### LINKED ISSUE #191499: [Inductor][CPU] torch.compile returns 0 for short positive uint8/int8 amin reductions on Apple Silicon
### 🐛 Describe the bug

On an Apple Silicon CPU, Inductor returns `0` for a short positive `uint8` or `int8` `amin` reduction, while eager execution returns the correct minimum. This is a silent wrong-result issue and reproduces on the first compiled call with a fresh Inductor cache.

The problem occurs with both static and dynamic compilation. A four-element tensor is sufficient; lengths 1–3 pass in my local checks, while lengths 4 and above fail.

### Minimal reproducer

```python
import torch

print("torch:", torch.__version__)
print("git:", torch.version.git_version)

for dtype in (torch.uint8, torch.int8):
    x = torch.tensor([5, 6, 7, 8], dtype=dtype)

    for dynamic in (False, True):
        torch._dynamo.reset()
        compiled_fn = torch.compile(
            lambda value: torch.amin(value),
            backend="inductor",
            fullgraph=True,
            dynamic=dynamic,
        )

        eager = torch.amin(x)
        compiled = compiled_fn(x)
        print(dtype, "dynamic=", dynamic,
              "eager=", eager.item(),
              "compiled=", compiled.item())

        torch.testing.assert_close(compiled, eager, rtol=0, atol=0)
```

Observed on PyTorch 2.13.0:

```text
torch.uint8 dynamic=False eager=5 compiled=0
torch.uint8 dynamic=True  eager=5 compiled=0
torch.int8  dynamic=False eager=5 compiled=0
torch.int8  dynamic=True  eager=5 compiled=0
```

Expected: all compiled results are `5`.

### Isolation results

- Reproduced 3/3 with a new `TORCHINDUCTOR_CACHE_DIR` each time.
- `uint8` and `int8` fail; `int16`, `int32`, `int64`, `float16`, and `float32` pass for the same values.
- Setting `torch._inductor.config.cpp.simdlen = 1` makes the `uint8` and `int8` cases correct.
- PyTorch 2.5.1 on the same machine returns the correct result for this reproducer.

The failing generated C++ uses a partial vector load followed by an unmasked vector minimum and horizontal minimum, approximately:

```cpp
auto tmp0 = at::vec::Vectorized<uint8_t>::loadu(in_ptr0, 4);
tmp_acc0_vec = at::vec::minimum(tmp_acc0_vec, tmp0);
tmp_acc0 = at::vec::vec_reduce_all<uint8_t, 1>(
    [](auto& x, auto& y) { return at::vec::minimum(x, y); },
    tmp_acc0_vec);
```

A possible explanation is that inactive lanes from the partial vector load participate as zero in the horizontal minimum. This is only a code-generation hint; the reproducer and `simdlen=1` control are the confirmed observations.

### Environment

```text
PyTorch: 2.13.0
PyTorch git: cf30153c4c131c8164ee7798e5022d810682e2cb
Python: 3.12.10
OS: macOS 26.0 arm64
Machine: Apple M1 Pro / arm64
Backend: Inductor CPU
```

### PR DESCRIPTION
**Pull request by JaynouOliver**
Fixes #191499

> [!NOTE]
> This PR (code and description) was authored with the assistance of an AI agent (Claude Code, model Claude Fable 5), per the repo [AI policy](https://github.com/pytorch/pytorch/blob/main/AI_POLICY.md). Opened as a draft pending human review.

The following analysis and fix description are AI-generated:

> ### Root cause
>
> For dtypes narrower than the vector lane width implied by the tiling factor (`uint8`/`int8` with a float-sized tiling factor), `CppVecKernel` loads fill only the first `tiling_factor` lanes of a `Vectorized<T>` and zero the rest, e.g. on NEON (tiling factor 4, 16 `uint8` lanes):
>
> ```cpp
> auto tmp0 = at::vec::Vectorized<uint8_t>::loadu(in_ptr0 + x0, 4);   // lanes 4..15 = 0
> tmp_acc0_vec = at::vec::minimum(tmp_acc0_vec, tmp0);                 // unmasked: identity lanes -> 0
> ...
> tmp_acc0 = min_propagate_nan(tmp_acc0, at::vec::vec_reduce_all<uint8_t, 1>(..., tmp_acc0_vec));  // reduces ALL 16 lanes
> ```
>
> The accumulator is initialized to the reduction identity in all lanes, but the unmasked combine in the main loop overwrites the inactive lanes (for `min`, `minimum(identity, 0) == 0`), and the horizontal `vec_reduce_all` in the reduction suffix then reduces across all lanes, leaking the polluted lanes into the result. So `torch.amin` of an all-positive `uint8` tensor compiles to `0`, and `torch.amax` of an all-negative `int8` tensor compiles to `0`, at any tensor length. The bug is ISA-independent (the tail loop is not the trigger; the main loop's partial loads are), and also affects `torch.min`/`torch.max`/`torch.aminmax`, and, through the same polluted value lanes, `argmin`/`argmax` (plus the indices of `torch.min/max(dim=)`). Reductions that promote (`sum`/`prod` to `int64`) compute in fully-utilized wider vectors and are unaffected; `bf16`/`fp16` compute in `float` and are likewise unaffected.
>
> ### Fix
>
> Two small pieces, one per reduction flavor:
>
> **min/max:** blend the inactive accumulator lanes back to the reduction identity right before the horizontal reduction, in the reduction suffix:
>
> ```cpp
> at::vec::vec_reduce_all<uint8_t, 1>(..., at::vec::Vectorized<uint8_t>::set(at::vec::Vectorized<uint8_t>(std::numeric_limits<uint8_t>::max()), tmp_acc0_vec, 4));
> ```
>
> gated on `_get_raw_num_vectors(dtype) < 1` (lane underutilization), mirroring the existing precedent in `CppVecOverrides.remainder`. The blend count is the tiling factor: the lanes past it are the ones the loads never fill. The alternative of masking every combine in the main loop (as the tail loop already does via `min_masked_reduce`) would cost a blend per loop iteration instead of a single blend in the suffix.
>
> **argmin/argmax:** `argmin/argmax_vec_reduce_all` iterated over the full value-vector width, reading polluted value lanes and index lanes that do not exist. Fixed entirely in `cpp_prefix.h`: the index vector always has exactly one lane per active element (`VectorizedN<int64_t, NI>::size()` equals the tiling factor on every ISA), so the horizontal reduce now derives its length from the index vector width instead of the value vector width. No codegen change; the emitted call is unchanged from upstream.
>
> ### Test plan
>
> New regression test covering `amin`/`amax`/`aminmax`/`argmin`/`argmax`/`min(dim=)`/`max(dim=)` over `uint8`/`int8`, swept across `simd_lengths_to_test()` and sizes hitting tail-only/main-only/main+tail loops, asserting the kernel vectorizes and FileChecking the generated identity blend:
>
> ```
> python -m pytest test/inductor/test_cpu_repro.py -k test_int8_min_max_reduction
> ```
>
> Verified on an Apple M-series machine (Inductor CPU, NEON) by patching the fix into the torch 2.14.0.dev20260703 nightly wheel: the issue reproducer returns 5 for all four variants; a sweep of `amin`/`amax`/`min`/`max`/`aminmax`/`sum`/`prod` (1D and 2D dims, static and dynamic shapes, sizes 4-100) over `uint8`/`int8`/`int16`/`int32`/`float32`/`bfloat16`/`float16` matches eager (359 checks); a 296-case argmin/argmax unique-extremum stress test and a 272-case tie/duplicated-extremum suite (inputs made entirely of the reduction identity, the adversarial case for tail-only kernels) all match eager. The AVX512/AVX2 template instantiations were syntax-checked via cross-compile.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: torch/csrc/inductor/cpp_prefix.h [hunk 1]
```diff
@@ -664,31 +664,37 @@ inline IndexValueVec<T, NV, NI>& argmax_combine_vec(
   return argmax_vec_impl(a, next_value, next_index, tail_size);
 }
 
+// For dtypes narrower than the vector lane width, only the first `len` value
+// lanes are active and the index vector has fewer lanes than the value vector,
+// so reducing the full width would read inactive values and uninitialized
+// indices.
 template <typename T, int NV, int NI>
 inline IndexValue<T> argmin_vec_reduce_all(
```
Comment (jananisriram) [inline torch/csrc/inductor/cpp_prefix.h:672]: from my agent:

The PR adds a 12-line gate in cpp.py that computes a dtype, tests _get_raw_num_vectors(val_dtype) < 1, and conditionally emits a second argument. All of that is unnecessary, because the header already knows the answer:
`VectorizedN<int64_t, NI>::size() == tiling_factor == num_elems` for every ISA and both tiling-factor paths (NI = ceil(tf*64 / bw), so NI*(bw / 64) == tf for tf ∈ {bw/32, bw/16}).

I implemented and tested the alternative:
```
inline IndexValue<T> argmin_vec_reduce_all(const IndexValueVec<T, NV, NI>& vec) {
  constexpr int max_len = at::vec::VectorizedN<T, NV>::size();
  constexpr int len = at::vec::VectorizedN<int64_t, NI>::size();   // the entire fix
  __at_align__ T tmpval[max_len];
  __at_align__ int64_t tmpidx[len];                                 // 512B -> 128B for int8/AVX512
  ...
  for (int i = 1; i < len; i++) ...
```
…with the whole cpp.py argmin hunk reverted to upstream. Result: 0 miscompiles across the same 2,940 checks (run with TORCHINDUCTOR_FORCE_DISABLE_CACHES=1, a stricter bar than the PR-as-written run), and codegen goes back to emitting the
plain one-argument argmin_vec_reduce_all(tmp_acc0_vec) — zero two-arg calls.

File: test/inductor/test_cpu_repro.py [hunk 2]
```diff
@@ -7413,6 +7413,37 @@ def forward(self, x):
                 # inf or finite numbers, it shouldn't degrade into unexpected NaNs).
                 self.assertEqual(eager_out, compiled_out)
 
+    @requires_vectorization
```
Comment (jananisriram) [inline test/inductor/test_cpu_repro.py:7417]: from my agent:

The new test exercises AVX2 (tiling factor 8) with TORCHINDUCTOR_CPP_WRAPPER=1, and nothing else. AVX512 (factor 16), the AMX runners, and the plain Python-wrapper path get zero coverage. I suggest a FileCheck on the generated C++ (`run_and_get_cpp_code`, precedent at `test_cpu_repro.py:5915`) asserting `::set(...)` and `argmin_vec_reduce_all(`, or parametrize over the existing `simd_lengths_to_test()` helper (`:100-106`).

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -3451,7 +3451,19 @@ def reduction(self, dtype, src_dtype, reduction_type, value):
...
+                val_dtype = (
+                    torch.float
+                    if src_dtype == torch.bool
+                    else DTYPE_TO_COMPUTATION_DTYPE[src_dtype]
+                )
+                if self._get_raw_num_vectors(val_dtype) < 1:
+                    # Partial loads leave the lanes past num_elems inactive; keep
+                    # them out of the horizontal reduction (see the min/max
+                    # branch below).
+                    next_value = f"{reduction_type}_vec_reduce_all({acc_vec}, {cexpr_index(self.num_elems)})"
```
Comment (jananisriram) [inline torch/_inductor/codegen/cpp.py]: from my agent:

`num_elems` is a no-op today, but it's the value that would be wrong if a tail kernel's suffix ever did get emitted (it would wipe main-loop results in lanes `[tail_size, tiling_factor)`), while `tiling_factor` would still be right. note that this reverses the earlier bot review, which pushed `tiling_factor → num_elems` on "mirror the load boundary" ground

File: torch/_inductor/codegen/cpp.py [hunk 4]
```diff
@@ -3479,6 +3491,15 @@ def reduction(self, dtype, src_dtype, reduction_type, value):
                     if reduction_type != "sum":
                         raise AssertionError('expected reduction_type == "sum"')
                     result_vec = f"{acc_vec} + {masked_acc_vec}"
+                if self._get_raw_num_vectors(vec_dtype) < 1:
+                    # Partial loads fill only num_elems lanes and zero the rest,
+                    # while vec_reduce_all reduces all lanes, so reset the
+                    # inactive lanes to the reduction identity first.
+                    result_vec = (
+                        f"{acc_type_vec}::set("
```
Comment (jananisriram) [inline torch/_inductor/codegen/cpp.py:3487]: from my agent:

this file now has 4 call sites that compensate for inactivate lanes in narrow dtypes. perhaps a one-line helper on CppVecKernel would single-source gate + blend across all four

Image

**Review by jananisriram — COMMENTED** [review]
left some comments! i'm not an inductor CPU expert, so might be worth flagging to someone who may know more about this.

**Comment by slayton58** [comment]
Perhaps this is a silly question, but why not do a serial "remainder" loop for the few non-vectorizable elements? I.e. if we have 102 elements, and vector-size == 4, we do 25 vectorized ops, then 2 serial ones at the end to handle the tail. IMO that feels cleaner & less error-prone than this

File: torch/csrc/inductor/cpp_prefix.h (hunk 1 above)
Comment (JaynouOliver) [reply torch/csrc/inductor/cpp_prefix.h:672]: From my agent:

> Adopted — `argmin/argmax_vec_reduce_all` now derive the reduce length as `constexpr int len = at::vec::VectorizedN<int64_t, NI>::size()` (with a `static_assert(len <= max_len)`), `tmpidx` shrinks to `len`, and the cpp.py argmin hunk is reverted to the upstream one-liner. One note on the tail-only kernel case: lanes in `[tail_size, len)` there hold the init state (identity value, index 0). They can only tie a real candidate, and a tie resolves to index 0, which is then the true arg-extremum, so reducing over the full index width stays correct. Added a 272-case tie/duplicated-extremum suite (inputs made entirely of the identity value) to cover exactly that, and reran the sweep and stress tests on NEON — all clean. AVX512/AVX2 instantiations syntax-checked via cross-compile.

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (JaynouOliver) [reply torch/_inductor/codegen/cpp.py]: From my agent:

> Adopted — the blend count is now `tiling_factor`, with the comment updated to state the invariant it mirrors (loads fill at most `tiling_factor` lanes; everything past that must return to the identity before `vec_reduce_all`).

File: test/inductor/test_cpu_repro.py (hunk 2 above)
Comment (JaynouOliver) [reply test/inductor/test_cpu_repro.py:7417]: From my agent:

> Adopted — the test now sweeps `simd_lengths_to_test()` (None, 1, and the primary ISA bit width) and FileChecks the generated C++ for the dtype-specific `int8_t>::set(`/`uint8_t>::set(` blend via `run_and_get_cpp_code`. I didn't FileCheck `argmin_vec_reduce_all(`: after the header-only change the emitted call is byte-identical to upstream's, so its presence no longer distinguishes the fix — the argmin path is covered by numerics (including the new tie cases, where the inactive lanes tie the real candidates). It would also pin the vectorized-argmin path against silent scalar fallback, but I can't confirm int8 argmin vectorizes on every CI ISA (AVX2 in particular), so I left it out rather than risk a false CI failure — happy to add it if you can confirm.

File: torch/_inductor/codegen/cpp.py (hunk 4 above)
Comment (JaynouOliver) [reply torch/_inductor/codegen/cpp.py:3487]: From my agent:

> After reverting the argmin hunk only two gated sites remain — `CppVecOverrides.remainder` and the reduction-suffix blend — and their compensations are structurally different (`blend<compile-time mask>(_t(1), b)` vs `::set(identity, acc, tiling_factor)`), so a shared helper would only centralize the one-line `_get_raw_num_vectors(dtype) < 1` predicate. Left as-is to keep the diff minimal, but happy to add a small `_has_inactive_lanes(dtype)` predicate if you think it carries its weight.

**Comment by JaynouOliver** [comment]
@slayton58 Not a silly question at all. From my agent:

> The tail isn't the trigger, so a serial remainder loop wouldn't remove the hazard. The kernel's vector width (tiling factor) is chosen per loop nest in float lanes — 4 on NEON, 16 on AVX512 — while `Vectorized<uint8_t>` physically holds 16/64 lanes. So in your example every one of the 25 "full" vector ops already loads only 4 of 16 lanes and zeroes the rest; the dead lanes get polluted on each main-loop iteration, and the final horizontal reduce reads all 16. The 2 tail elements are handled with masked ops today and were never the problem. Making the main loop fill all 16 lanes would need a dtype-dependent tiling factor, but the factor is shared by every op fused into the loop nest, so that's a much larger change to the tiling model. Given that, resetting the inactive lanes once in the reduction suffix (and, after the latest push, deriving the argmin/argmax reduce length from the index-vector width in the header) is the minimal change consistent with the existing codegen.
